source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,416 @@
+}
+
+/// Logger instance
+private Logger log;
+
+/*******************************************************************************
+
+    Handles data and process for enrollment
+
+*******************************************************************************/
+
+public class EnrollmentManager
+{
+    /// SQLite db instance
+    private Database db;
+
+    /// Node's key pair
+    private KeyPair node_keypair;
+
+    /// Random seed
+    private Scalar random_seed;
+
+    /// Path to random seed file on disk
+    private string random_seed_path;
+
+    /// Random key for enrollment
+    private Pair schnorr_pair;
+
+    /// EnrollmentData
+    private EnrollmentData* data;
+
+    /// Database path
+    private string db_path;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
+            keypair = the keypair of the owner node
+
+    ***************************************************************************/
+
+    public this (in string db_path, in string seed_file_path, KeyPair keypair)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
+
+        this.db = Database(db_path);
+
+        // create the table for validator set if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS validator_set " ~
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+
+        // load random seed
+        this.random_seed_path = seed_file_path;
+        if (this.random_seed_path.exists())
+        {
+            auto seed_file = File(this.random_seed_path, "r");
+            scope DeserializeDg dg = (size) @safe
+            {
+                ubyte[] res;
+                res.length = size;
+                seed_file.rawRead(res);
+                return res;
+            };
+
+            this.random_seed.deserialize(dg);
+        }
+
+        // set information to member variables
+        this.db_path = db_path;
+        this.node_keypair = keypair;
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a enrollment data to the set
+
+        Params:
+            enroll_data = the enrollment data to add
+
+        Returns:
+            true if the enrollment data has been added to the set
+
+    ***************************************************************************/
+
+    public bool add (EnrollmentData enroll_data) @safe
```

Comment:
This function unconditionally returns `true`. Maybe you could check `hasEnrollmentDataHash` and return `false` early if `hasEnrollmentDataHash` return true?